frontend/controller/actions/group.js

```diff
@@ -131,6 +181,101 @@ export default (sbp('sbp/selectors/register', {
   'gi.actions/group/switch': function (groupId) {
     sbp('state/vuex/commit', 'setCurrentGroupId', groupId)
   },
+  'gi.actions/group/addChatRoom': async function (params: GIActionParams) {
+    const message = await sbp('gi.actions/chatroom/create', { data: params.data })
+
+    await sbp('chelonia/out/actionEncrypted', {
+      ...params,
+      action: 'gi.contracts/group/addChatRoom',
+      data: {
+        ...params.data,
+        chatRoomID: message.contractID()
+      }
+    })
+
+    return message
+  },
+  'gi.actions/group/joinChatRoom': async function (params: GIActionParams) {
+    try {
+      const rootState = sbp('state/vuex/state')
+      const username = params.data.username || rootState.loggedIn.username
+      await sbp('gi.actions/chatroom/join', {
+        contractID: params.data.chatRoomID,
+        data: { username }
+      })
+
+      if (username === rootState.loggedIn.username) {
+        sbp('okTurtles.data/set', 'READY_TO_JOIN_CHATROOM', true)
+      }
+      await sbp('chelonia/out/actionEncrypted', {
+        ...params,
+        action: 'gi.contracts/group/joinChatRoom'
+      })
+      sbp('okTurtles.data/set', 'READY_TO_JOIN_CHATROOM', false)
```

Comment:
It doesn't make sense to surround `await sbp('chelonia/out/actionEncrypted'` - because that will only wait for the message to be sent - not for it to be fully processed.

Both of the calls to `'okTurtles.data/set'` should be removed, and it shouldn't cause any problems. If, after applying all the other changes in this review, it does cause problems (and you tested it to verify), then please describe the scenario that breaks and I will try testing it myself.

Reply:
Correct. Really nice catch. We definitely need `READ_TO_JOIN_CHATROOM`.
So I wanna call this `sbp('okTurtles.data/set', 'READY_TO_JOIN_CHATROOM', false)` at the end of the `sideEffect` of `gi.contracts/group/joinChatRoom`.
Just like the following.
```

async sideEffect ({ meta, data }, { state }) {
        const rootState = sbp('state/vuex/state')
        const username = data.username || meta.username
        if (username === rootState.loggedIn.username) {
          if (!sbp('okTurtles.data/get', 'JOINING_GROUP') || sbp('okTurtles.data/get', 'READY_TO_JOIN_CHATROOM')) {
            // while users are joining chatroom, they don't need to leave chatrooms
            // this is similar to setting 'JOINING_GROUP' before joining group
            sbp('okTurtles.data/set', 'JOINING_CHATROOM', true)
            await sbp('gi.actions/contract/syncAndWait', data.chatRoomID)
            sbp('okTurtles.data/set', 'JOINING_CHATROOM', false)
            sbp('okTurtles.data/set', 'READY_TO_JOIN_CHATROOM', false)
          }
        }
      }
```